### ISSUE TITLE
Move to Erlang + Rebar3

### ISSUE THREAD

**Issue by josevalim** [beam-telemetry/telemetry#20]
This will allow this package to be used by everyone in the community for event dispatching, which will improve the coverage of events throughout the ecosystem. Right now we are planning to move only the core though.

@tsloughter, can you please review the current API and let us know if you would like to do any changes in the process?

**Comment by josevalim** [comment]
We need to figure out the docs though. Suggestions?

**Comment by josevalim** [comment]
We should also rename the GitHub organization most likely.

**Comment by chrismccord** [comment]
erlang-telemetry, beam-telemetry ?

**Comment by OvermindDL1** [comment]
Isn't it fairly easy to make an erlang API in elixir though?  Is this being talked about to change the entire language of the project?

**Comment by josevalim** [comment]
This is supposed to be a very tiny dependency, so if we feel that if developers are forced to bring elixir/mix only to publish the events, it won't have the same effect in the Erlang community. So this would be a complete move to Erlang.

**Comment by tsloughter** [comment]
@josevalim Will do this morning.

For "figure out the docs" do you mean because edoc is ugly? :). I have some hacks I do for opencensus and pretty sure I did for another Erlang project but I forget which... that make them more modern https://hexdocs.pm/opencensus/

**Comment by wojtekmach** [comment]
Here’s an example PR for that approach https://github.com/hexpm/hex_core/pull/39 (pretty much ready, blocked on me on minor stuff)

**Comment by arkgil** [comment]
Regarding organization name, I like beam-telemetry. Opencensus docs look really nice, let's go for this approach! :)

**Comment by tsloughter** [comment]
Ah right, it was hex_core, thanks @wojtekmach :)

**Comment by tsloughter** [comment]
Right now for the API I'll just say what it would have looked like had I done it from scratch. Reason I'm doing that is I can't decide if any of it is useful or if it all works fine anyway because the values still "fit" and pass through (like if the event name is a list of atoms as it is now and I need just an atom it would just be I match out the first element).

```
-spec execute(Ctx :: ctx:t() | MetaData :: map(), EventId :: atom(), Value :: term()).

-spec execute(Ctx :: ctx:t() | MetaData :: map(), [{EventId :: atom(), Value :: term()}]).
```

The main thing here is the use of a generic context argument (or a map of metadata like the current API) so that tags/labels don't have to be pulled out from the context every time you call `execute`. And I think pushing a generic context library that supports carrying the metadata and stuff like deadlines is useful. I'm not tied to mine https://github.com/tsloughter/ctx (though I use it in grpcbox currently), it is simple library and happy to have a completely different one started if people are interested in such a context library we could all agree on.

I would also have a callback for when `attach` is called. This may mean adding a metadata argument, depending on what you were thinking the `config` argument to attach was for. I see using `attach` in the case of OpenCensus being used to call view subscribe or measure registration and those require descriptions, units and tags/labels to go along with the view or measure name.

**Comment by josevalim** [comment]
@tsloughter regarding the execute call itself, I believe the APIs are equivalent. We don't support multiple events at once but that can be easily done by the caller too.

Regarding the context, could you store the context inside the metadata map? I believe forcing a map at the top level is better because it makes it easier to tag/filter events later on. For example, in Phoenix I would be able to do:

    histogram([:my_app, :slow_controller, :render], filters: [controller: MyApp.SlowController], unit: :us)

And that would only consider events that have `controller: MyApp.SlowController` in the metadata. Or I could say I want to push the controller and action as tags to the backend:

    counter([:my_app, :slow_controller, :render], tags: [:controller, :action], unit: :us)

If we allow the map to be anything, then the functionality above would work only part of the time. However, you should be able to pass any context inside the metadata. Does this make sense?

> I see using attach in the case of OpenCensus being used to call view subscribe or measure registration and those require descriptions, units and tags/labels to go along with the view or measure name.

The way we are tackling this right now is that metrics are completely decoupled from this project, as in this example above. All that this cares about is emitting the events. What those events mean, their units and description actually belong elsewhere. So in the example above, we had this:

    counter([:my_app, :slow_controller, :render], tags: [:controller, :action], unit: :us)

So something will receive the metric above (which is a separate package and not related to this package) that describes how the event is aggregated and what it means and what should be published and attach a handler for it. There is a bit of duplication here but we wanted to keep this part of the code decoupled from all of the rest on purpose, precisely so we focus only on the events themselves.

Thoughts?

**Comment by josevalim** [comment]
I had a brief conversation with @tsloughter and team and I think we are good to go with the current API, we just need to move it to Erlang. The context stuff should be addressable on top of the existing API. I will rename the organization to beam-telemetry. :)

**Comment by tsloughter** [comment]
@josevalim I can take a crack at the converting to Erlang. Just wanted to make sure someone hadn't already done so before I start.

**Comment by josevalim** [comment]
@arkgil did you start it or can @tsloughter go ahead? :D

**Comment by arkgil** [comment]
Yes, @tsloughter, thanks! 😊
